File: sdks/csharp/src/DocDigitizer/Exceptions.cs
namespace DocDigitizer;

/// <summary>Base exception for all DocDigitizer SDK errors.</summary>
public class DocDigitizerException : Exception
{
    public int? StatusCode { get; }
    public string? TraceId { get; }
    public IReadOnlyList<string>? Messages { get; }
    public IDictionary<string, object>? Timers { get; }

    public DocDigitizerException(string message, int? statusCode = null, string? traceId = null, IReadOnlyList<string>? messages = null, IDictionary<string, object>? timers = null)
        : base(message)
    {
        StatusCode = statusCode;
        TraceId = traceId;
        Messages = messages;
        Timers = timers;
    }
}

/// <summary>HTTP 401 — Invalid or missing API key.</summary>
public class AuthenticationException : DocDigitizerException
{
    public AuthenticationException(string message, int? statusCode = null, string? traceId = null, IReadOnlyList<string>? messages = null, IDictionary<string, object>? timers = null)
        : base(message, statusCode, traceId, messages, timers) { }
}

/// <summary>HTTP 400 — Invalid request.</summary>
public class ValidationException : DocDigitizerException
{
    public ValidationException(string message, int? statusCode = null, string? traceId = null, IReadOnlyList<string>? messages = null, IDictionary<string, object>? timers = null)
        : base(message, statusCode, traceId, messages, timers) { }
}

/// <summary>HTTP 429 — Rate limit exceeded.</summary>
public class RateLimitException : DocDigitizerException
{
    public RateLimitException(string message, int? statusCode = null, string? traceId = null, IReadOnlyList<string>? messages = null, IDictionary<string, object>? timers = null)
        : base(message, statusCode, traceId, messages, timers) { }
}

/// <summary>HTTP 500 — Internal server error.</summary>
public class ServerException : DocDigitizerException
{
    public ServerException(string message, int? statusCode = null, string? traceId = null, IReadOnlyList<string>? messages = null, IDictionary<string, object>? timers = null)
        : base(message, statusCode, traceId, messages, timers) { }
}

/// <summary>HTTP 503 — Downstream service unavailable.</summary>
public class ServiceUnavailableException : DocDigitizerException
{
    public ServiceUnavailableException(string message, int? statusCode = null, string? traceId = null, IReadOnlyList<string>? messages = null, IDictionary<string, object>? timers = null)
        : base(message, statusCode, traceId, messages, timers) { }
}

/// <summary>HTTP 504 — Processing timeout.</summary>
public class TimeoutException : DocDigitizerException
{
    public TimeoutException(string message, int? statusCode = null, string? traceId = null, IReadOnlyList<string>? messages = null, IDictionary<string, object>? timers = null)
        : base(message, statusCode, traceId, messages, timers) { }
}

public static class ErrorMapping
{
    private static readonly Dictionary<int, Func<string, int?, string?, IReadOnlyList<string>?, IDictionary<string, object>?, DocDigitizerException>> Map = new()
    {
        { 400, (m, s, t, msgs, tmrs) => new ValidationException(m, s, t, msgs, tmrs) },
        { 401, (m, s, t, msgs, tmrs) => new AuthenticationException(m, s, t, msgs, tmrs) },
        { 429, (m, s, t, msgs, tmrs) => new RateLimitException(m, s, t, msgs, tmrs) },
        { 500, (m, s, t, msgs, tmrs) => new ServerException(m, s, t, msgs, tmrs) },
        { 503, (m, s, t, msgs, tmrs) => new ServiceUnavailableException(m, s, t, msgs, tmrs) },
        { 504, (m, s, t, msgs, tmrs) => new TimeoutException(m, s, t, msgs, tmrs) },
    };

    internal static readonly HashSet<int> RetryableStatusCodes = new() { 429, 500, 503, 504 };

    public static DocDigitizerException RaiseForStatus(int statusCode, string? traceId = null, IReadOnlyList<string>? messages = null, IDictionary<string, object>? timers = null)
    {
        var msg = messages != null && messages.Count > 0
            ? string.Join("; ", messages)
            : $"HTTP {statusCode}";

        if (Map.TryGetValue(statusCode, out var factory))
            return factory(msg, statusCode, traceId, messages, timers);

        return new DocDigitizerException(msg, statusCode, traceId, messages, timers);
    }
}

File: sdks/python/src/docdigitizer/exceptions.py
"""DocDigitizer exception hierarchy.

Maps HTTP status codes to specific exception types for structured error handling.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional


class DocDigitizerError(Exception):
    """Base exception for all DocDigitizer SDK errors."""

    def __init__(
        self,
        message: str,
        *,
        status_code: Optional[int] = None,
        trace_id: Optional[str] = None,
        messages: Optional[List[str]] = None,
        timers: Optional[Dict[str, Any]] = None,
    ) -> None:
        self.status_code = status_code
        self.trace_id = trace_id
        self.messages = messages or []
        self.timers = timers
        super().__init__(message)


class AuthenticationError(DocDigitizerError):
    """Raised when the API key is invalid or missing (HTTP 401)."""

    pass


class ValidationError(DocDigitizerError):
    """Raised when the request is invalid (HTTP 400).

    Common causes: missing required fields, invalid PDF, bad UUID format.
    """

    pass


class RateLimitError(DocDigitizerError):
    """Raised when rate limits are exceeded (HTTP 429)."""

    pass


class ServerError(DocDigitizerError):
    """Raised when the server encounters an internal error (HTTP 500)."""

    pass


class ServiceUnavailableError(DocDigitizerError):
    """Raised when a downstream service is unreachable (HTTP 503)."""

    pass


class TimeoutError(DocDigitizerError):
    """Raised when processing takes too long (HTTP 504)."""

    pass


_STATUS_CODE_MAP = {
    400: ValidationError,
    401: AuthenticationError,
    429: RateLimitError,
    500: ServerError,
    503: ServiceUnavailableError,
    504: TimeoutError,
}


def raise_for_status(
    status_code: int,
    *,
    trace_id: Optional[str] = None,
    messages: Optional[List[str]] = None,
    timers: Optional[Dict[str, Any]] = None,
) -> None:
    """Raise the appropriate exception for a non-2xx status code."""
    if status_code < 400:
        return

    exc_class = _STATUS_CODE_MAP.get(status_code, DocDigitizerError)
    msg = "; ".join(messages) if messages else f"HTTP {status_code}"
    raise exc_class(
        msg,
        status_code=status_code,
        trace_id=trace_id,
        messages=messages,
        timers=timers,
    )
